jwst/stpipe/utilities.py

```diff
@@ -144,3 +149,63 @@ def folder_traverse(folder_path, basename_regex='.+', path_exclude_regex='^$'):
         for file in files:
             if basename_regex.match(file):
                 yield os.path.join(root, file)
+
+
+def record_step_status(datamodel, cal_step, success=True):
+    """Record whether or not a step completed in meta.cal_step
+
+    Parameters
+    ----------
+    datamodel : `~jwst.datamodels.JwstDataModel` instance
+        This is the datamodel or container of datamodels to modify in place
+
+    cal_step : str
+        The attribute in meta.cal_step for recording the status of the step
+
+    success : bool
+        If True, then 'COMPLETE' is recorded.  If False, then 'SKIPPED'
+    """
+    if success:
+        status = 'COMPLETE'
+    else:
+        status = 'SKIPPED'
+
+    if isinstance(datamodel, Sequence):
+        for model in datamodel:
+            model.meta.cal_step._instance[cal_step] = status
+    else:
+        datamodel.meta.cal_step._instance[cal_step] = status
+
+    # TODO: standardize cal_step naming to point to the official step name
+
+
+def query_step_status(datamodel, cal_step):
+    """Query the status of a step in meta.cal_step
+    
+    Parameters
+    ----------
+    datamodel : `~jwst.datamodels.JwstDataModel` instance
+        This is the datamodel or container of datamodels to check
+        
+    cal_step : str
+        The attribute in meta.cal_step to check
+    
+    Returns
+    -------
+    status : str
+        The status of the step in meta.cal_step, typically 'COMPLETE' or 'SKIPPED'
+    """
+    if isinstance(datamodel, Sequence):
+        status = np.array([getattr(model.meta.cal_step, cal_step, NOT_SET) for model in datamodel])
+        if np.any(status == COMPLETE):
+            return COMPLETE
+        elif np.all(status == NOT_SET):
+            return NOT_SET
+        elif np.all(status == SKIPPED):
```

Comment:
I think this could be `any` instead of `all` -- say, a step encountered a condition that was skippable in the first model and didn't end up setting a status for the remaining models.

Reply:
Or maybe that's just bad behavior on the step's part, and it should throw the error?

Reply:
As we agreed in stand-up, posting here for anyone else interested: if a condition encountered in the first model makes it so you want to skip all the models, then all the models should have their status set to skipped.

Reply:
Does that mean the `all` and `any` can be dropped and just the first model of the container inspected?

Reply:
No, because it's still an acceptable condition for the step to be complete on one model but skipped in others.

Reply:
Thanks. This seems to be a bit more involved than what I'd expect in a utility function. It seems like there are a bunch of hidden rules baked into the logic here which are in part expectations about how the steps set the status of the models in the container. Are these expectations documented and tested?

What if this function instead returned a list of statuses and left it up to the step to determine what is "COMPLETE", "SKIPPED", etc?

For the uses of this function in this PR I think returning a list should be sufficient and more flexible than having this function perform the combination logic. I'll add some comments to the usage to try to explain the idea.